File: day05/Part1.hs
import Data.Char (toLower)
import Data.List (isInfixOf, tails)

main :: IO ()
main = readFile "./input.txt" >>= print . length . filter isNice . words

isNice :: String -> Bool
isNice s =
  let vowels_count = length $ filter (`elem` "aeiou") s
      has_duplicate = or $ zipWith (==) s (tail s)
      is_banned = any (`isInfixOf` s) ["ab", "cd", "pq", "xy"]
   in vowels_count >= 3 && has_duplicate && not is_banned

File: day05/part1.py
import operator

def is_nice(word: str) -> bool:
    # duplicates
    if not any(map(operator.eq, word, word[1:])):
        return False

    VOWELS = "aeiou"
    if len(list(filter(lambda c: c in VOWELS, word))) < 3:
        return False

    BANNED = ["ab", "cd", "pq", "xy"]
    if any(map(lambda w: w in word, BANNED)):
        return False

    return True

with open("./input.txt") as f:
    words = map(str.strip, f.readlines())

print(len(list(filter(is_nice, words))))
